cmd/uhc/cluster/login/login.go

```diff
@@ -152,3 +168,69 @@ func run(cmd *cobra.Command, argv []string) {
 		os.Exit(1)
 	}
 }
+
+// doSurvey will ask user to choose one if there are more than one clusters match the query
+func doSurvey(clusters []*v1.Cluster) (cluster *v1.Cluster, err error) {
+	clusterList := []string{}
+	for _, v := range clusters {
+		clusterList = append(clusterList, fmt.Sprintf("Name: %s, ID: %s", v.Name(), v.ID()))
+	}
+	choice := ""
+	prompt := &survey.Select{
+		Message: "Please choose a cluster:",
+		Options: clusterList,
+		Default: clusterList[0],
+	}
+	err = survey.AskOne(prompt, &choice, func(ans interface{}) error {
+		choice := ans.(string)
+		found := false
+		for _, v := range clusters {
+			if strings.Contains(choice, v.ID()) {
+				found = true
+				cluster = v
+			}
+		}
+		if !found {
+			return fmt.Errorf("the cluster you choose is not valid: %s", choice)
+		}
+		return nil
+	})
+	return cluster, err
+}
+
+// findClusters finds the clusters that match the given key. A cluster matches the key if its
+// identifier is that key, or if its name starts with that key. For example, the key `prd-2305`
+// doesn't match a cluster directly because it isn't a valid identifier, but it matches all clusters
+// whose names start with `prd-2305`.
+func findClusters(collection *v1.ClustersClient, key string) (clusters []*v1.Cluster, err error) {
+
+	// Get the resource that manages the cluster that we want to display:
+	clusterResource := collection.Cluster(key)
+	response, err := clusterResource.Get().Send()
+
+	// If it's not an cluster id, try to query clusters using search param, we assume the
+	// list should be within 50, if not, we need update the cluster names prefix to minimize
+	// the list
+	if err != nil {
```

Comment:
Here you are assuming that error was a `404 Not Found`, but it may be other thing. For example may be an authorization error, or a time out. To make this more robust you need to check explicitly those cases. Note also that when there is no response the returned `response` will be `nil`, so you have to be careful to not use it, or it will generate a panic. I'd suggest that you check those cases explicitly:

```go
// If there was no response, or if the error was something other
// than `404 Not Found` then the only thing we can do is report
// it back to the caller:
if response == nil || response.Status() != http.SatusNotFound {
        return
}
```